libafl_libfuzzer/libafl_libfuzzer_runtime/src/fuzz.rs

```diff
@@ -0,0 +1,224 @@
+    time::{SystemTime, UNIX_EPOCH},
+};
+#[cfg(unix)]
+use std::{
+    io::Write,
+    os::fd::{AsRawFd, FromRawFd, IntoRawFd},
+};
+
+use libafl::{
+    corpus::Corpus,
+    events::{
+        launcher::Launcher, EventConfig, ProgressReporter, SimpleEventManager,
+        SimpleRestartingEventManager,
+    },
+    executors::ExitKind,
+    inputs::UsesInput,
+    monitors::{
+        tui::{ui::TuiUI, TuiMonitor},
+        Monitor, MultiMonitor, SimpleMonitor,
+    },
+    stages::StagesTuple,
+    state::{
+        HasClientPerfMonitor, HasExecutions, HasLastReportTime, HasMetadata, HasSolutions,
+        UsesState,
+    },
+    Error, Fuzzer,
+};
+use libafl_bolts::{
+    core_affinity::Cores,
+    shmem::{ShMemProvider, StdShMemProvider},
+};
+
+use crate::{feedbacks::LibfuzzerCrashCauseMetadata, fuzz_with, options::LibfuzzerOptions};
+
+fn do_fuzz<F, ST, E, S, EM>(
+    options: &LibfuzzerOptions,
+    fuzzer: &mut F,
+    stages: &mut ST,
+    executor: &mut E,
+    state: &mut S,
+    mgr: &mut EM,
+) -> Result<(), Error>
+where
+    F: Fuzzer<E, EM, ST, State = S>,
+    S: HasClientPerfMonitor
+        + HasMetadata
+        + HasExecutions
+        + UsesInput
+        + HasSolutions
+        + HasLastReportTime,
+    E: UsesState<State = S>,
+    EM: ProgressReporter<State = S>,
+    ST: StagesTuple<E, EM, S, F>,
+{
+    if let Some(solution) = state.solutions().last() {
+        let kind = state
+            .solutions()
+            .get(solution)
+            .expect("Last solution was not available")
+            .borrow()
+            .metadata::<LibfuzzerCrashCauseMetadata>()
+            .expect("Crash cause not attached to solution")
+            .kind();
+        let mut halt = false;
+        match kind {
+            ExitKind::Oom if !options.ignore_ooms() => halt = true,
+            ExitKind::Crash if !options.ignore_crashes() => halt = true,
+            ExitKind::Timeout if !options.ignore_timeouts() => halt = true,
+            _ => {
+                log::info!("Ignoring {kind:?} according to requested ignore rules.");
+            }
+        }
+        if halt {
+            log::info!("Halting; the error on the next line is actually okay. :)");
+            return Err(Error::shutting_down());
+        }
+    }
+    fuzzer.fuzz_loop(stages, executor, state, mgr)?;
+    Ok(())
+}
+
+fn fuzz_single_forking<M>(
+    options: &LibfuzzerOptions,
+    harness: &extern "C" fn(*const u8, usize) -> c_int,
+    mut shmem_provider: StdShMemProvider,
+    monitor: M,
+) -> Result<(), Error>
+where
+    M: Monitor + Debug,
+{
+    fuzz_with!(options, harness, do_fuzz, |fuzz_single| {
+        let (state, mgr): (
+            Option<StdState<_, _, _, _>>,
+            SimpleRestartingEventManager<_, StdState<_, _, _, _>, _>,
+        ) = match SimpleRestartingEventManager::launch(monitor, &mut shmem_provider) {
+            // The restarting state will spawn the same process again as child, then restarted it each time it crashes.
```

Comment:
i guess if you launch the manager everytime, then the timestamp is not preserved across restarts

Reply:
Maybe fix this after merging/open an issue?